### PR TITLE
(fix) O3-2081- Update tests for patient-registration-validation component

### PR DESCRIPTION
**Pull request by ayush-AI**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

In this PR, I tried to update the failing tests which was getting skipped. The test was failing because the error text was not coming when input field was left empty. I tried to fix those by mocking the input component. I also mocked the formik component.

## Screenshots

Screenshot 2023-05-29 at 10 56 22 PM

*None.*

## Related Issue

https://issues.openmrs.org/browse/O3-2081

*None.*

## Other

Please carefully check If I am mocking input component and formik correctly or not. 

*None.*

### CODE REVIEW

**Comment by ayush-AI** [comment]
cc / @jayasanka-sack @anjula-sack @Piumal1999

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 1]
```diff
@@ -112,18 +210,19 @@ describe('Name input', () => {
...
     };
   };
 
   testValidName('Aaron', 'A', 'Aaronson');
-  testValidName('No', '', 'Middle Name');
-  testInvalidName('', '', '', 'Given name is required', 'givenNameError');
-  testInvalidName('', '', '', 'Family name is required', 'familyNameError');
-  testInvalidName('', 'No', 'Given Name', 'Given name is required', 'givenNameError');
-  testInvalidName('No', 'Family Name', '', 'Family name is required', 'familyNameError');
+  // testValidName('No', '', 'Middle Name');   // this test fails because the error message is showing up for the middle name
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:223]: What kind of error message does it show?

Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:223]: Middle Name is required is the error message but it shouldn't show because this field is optional.

Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx]: @Piumal1999 I fixed this test. Thank you. Can you please first review the mocks for `input` and `formik` components. I have doubt on them.

**Comment by Piumal1999** [comment]
The `patient-registration-validation.ts` contains only a `validationSchema` implemented using `Yup` for validating patient registration form. So I think, we should use the `patient-registration-validation.test.ts` specifically for **unit tests** related to the validation schema. There we can test different combinations of `givenName`, `middleName`,'familyName' (similar to what you have already done) to ensure the correct behaviour.

For integration tests, we can use a different directory outside of the current one. What do you think? @anjula-sack @jayasanka-sack @denniskigen

**Comment by ayush-AI** [comment]
> The `patient-registration-validation.ts` contains only a `validationSchema` implemented using `Yup` for validating patient registration form. So I think, we should use the `patient-registration-validation.test.ts` specifically for **unit tests** related to the validation schema. There we can test different combinations of `givenName`, `middleName`,'familyName' (similar to what you have already done) to ensure the correct behaviour.
> 
> For integration tests, we can use a different directory outside of the current one in future. What do you think? @anjula-sack @jayasanka-sack @denniskigen

should I go ahead and update the PR with the unit tests? @jayasanka-sack @Piumal1999 @denniskigen @anjula-sack

**Comment by Piumal1999** [comment]
> > The `patient-registration-validation.ts` contains only a `validationSchema` implemented using `Yup` for validating patient registration form. So I think, we should use the `patient-registration-validation.test.ts` specifically for **unit tests** related to the validation schema. There we can test different combinations of `givenName`, `middleName`,'familyName' (similar to what you have already done) to ensure the correct behaviour.
> > For integration tests, we can use a different directory outside of the current one in future. What do you think? @anjula-sack @jayasanka-sack @denniskigen
> 
> should I go ahead and update the PR with the unit tests? @jayasanka-sack @Piumal1999 @denniskigen @anjula-sack

Yeah. I think we can use this PR to write the unit tests for the validator. When writing integration tests, you can refer to the previous code.

**Comment by ayush-AI** [comment]
I updated the PR by changing the tests from integration to unit for the patient-registration-validation component. Please review @Piumal1999 @anjula-sack @jayasanka-sack

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 2]
```diff
@@ -1,129 +1,143 @@
...
+      const invalidFormValues = {
+        ...validFormValues,
+        birthdateEstimated: true,
+        yearsEstimated: 0,
+        monthsEstimated: -1,
+      };
+      const validationError = await validateFormValues(invalidFormValues);
+      expect(validationError.errors).toContain('negativeMonths');
+    });
+    it('should throw error when deathDate is not in future', async () => {
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:134]:
```suggestion
    });
    
    it('should throw error when deathDate is not in future', async () => {
```

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 3]
```diff
@@ -1,129 +1,143 @@
...
+      },
+    };
 
-  return {
-    ...originalModule,
-    useConfig: jest.fn(),
-    validator: jest.fn(),
-  };
-});
+    const validateFormValues = async (formValues) => {
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:27]: We can do something like this:
```
const validateFormValues = async (formValues) => {
  try {
    await validationSchema.validate(formValues, { abortEarly: false });
  } catch (err) {
    return err;
  }
};
```

**Review by Piumal1999 — CHANGES_REQUESTED** [review]
Good job @ayush-AI 
I added some comments. Also, let's reformat the code to make it more readable.

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx (hunk 3 above)
Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:27]: @Piumal1999 updated the PR please review

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 4]
```diff
@@ -1,129 +1,143 @@
...
+      expect(validationError).toBeFalsy();
+    });
 
-    fireEvent.change(givenNameInput, { target: { value: givenNameValue } });
-    fireEvent.blur(givenNameInput);
-    fireEvent.change(middleNameInput, { target: { value: middleNameValue } });
-    fireEvent.blur(middleNameInput);
-    fireEvent.change(familyNameInput, { target: { value: familyNameValue } });
-    fireEvent.blur(familyNameInput);
+    it('birthdays should not be in the future', async () => {
```
Comment (jayasanka-sack) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:105]: can you rephrase this?

**Comment by ayush-AI** [comment]
@jayasanka-sack @Piumal1999 @anjula-sack could you please check if anything more need to be changed ?

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 5]
```diff
@@ -1,129 +1,143 @@
...
+      expect(validationError.errors).toContain('negativeMonths');
+    });
 
-  testValidName('Aaron', 'A', 'Aaronson');
-  testValidName('No', '', 'Middle Name');
-  testInvalidName('', '', '', 'Given name is required', 'givenNameError');
-  testInvalidName('', '', '', 'Family name is required', 'familyNameError');
-  testInvalidName('', 'No', 'Given Name', 'Given name is required', 'givenNameError');
-  testInvalidName('No', 'Family Name', '', 'Family name is required', 'familyNameError');
+    it('should throw error when deathDate is not in future', async () => {
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:134]: Actually, it should throw an error if the deathDate **is a** future date. Can you check the validator and fix it?

cc: @vasharma05

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx [hunk 6]
```diff
@@ -1,129 +1,143 @@
...
+      expect(validationError).toBeFalsy();
+    });
 
-    fireEvent.change(givenNameInput, { target: { value: givenNameValue } });
-    fireEvent.blur(givenNameInput);
-    fireEvent.change(middleNameInput, { target: { value: middleNameValue } });
-    fireEvent.blur(middleNameInput);
-    fireEvent.change(familyNameInput, { target: { value: familyNameValue } });
-    fireEvent.blur(familyNameInput);
+    it('date of birth should not be in the future', async () => {
```
Comment (Piumal1999) [inline packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx:105]: Something like this?

``` suggestion
    it('should throw error when date of birth is a future date', async () => {
```

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx (hunk 5 above)
Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx]: @Piumal1999 the test and validator is fine. The issue is in the test name. I changed the date to a passed date and it was not throwing the error and test was failing due to that. I updated the test name.

File: packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx (hunk 6 above)
Comment (ayush-AI) [reply packages/esm-patient-registration-app/src/patient-registration/validation/patient-registration-validation.test.tsx]: updated this also. Thank you

**Review by Piumal1999 — APPROVED** [review]
Good job @ayush-AI 
Thank you
